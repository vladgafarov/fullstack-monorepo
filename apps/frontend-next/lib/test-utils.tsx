import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { Provider } from 'react-redux'
import { store } from '@redux/store'

const AllTheProviders = ({ children }) => {
   return <Provider store={store}>{children}</Provider>
}

const customRender = (
   ui: ReactElement,
   options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
