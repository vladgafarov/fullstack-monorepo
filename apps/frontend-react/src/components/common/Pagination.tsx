import { Button } from 'ui'
import { perPage } from '@lib/config'
import {
   BiChevronLeft,
   BiChevronRight,
   BiChevronsRight,
   BiChevronsLeft,
} from 'react-icons/bi'
import { useNavigate } from 'react-router-dom'

const Pagination = ({
   length,
   page,
   isFetching,
}: {
   length: number
   page: number
   isFetching?: boolean
}) => {
   const navigate = useNavigate()

   const handlePrevPage = () => {
      navigate(`?page=${page - 1}`)
   }

   const handleNextPage = () => {
      navigate(`?page=${page + 1}`)
   }

   const handleButtonClick = (page: number) => {
      navigate(`?page=${page}`)
   }

   const pages = Math.ceil(length / perPage)

   if (pages < 2) {
      return null
   }

   return (
      <>
         <div className="flex items-center justify-center mt-12">
            <button
               data-testid="prev"
               onClick={() => navigate(`?page=1`)}
               disabled={page === 1}
               className="text-3xl text-blue-500"
            >
               <BiChevronsLeft />
            </button>
            <button
               onClick={handlePrevPage}
               disabled={page === 1}
               className="text-3xl text-blue-500"
            >
               <BiChevronLeft />
            </button>
            <div className="space-x-2">
               {Array.from({ length: pages >= 5 ? 5 : pages }, (_, i) => {
                  const diff = page - 3
                  const diffEnd = pages - 2

                  if (pages >= 5) {
                     if (page === pages) i += page - 5
                     else if (page > diffEnd) i += page - 4
                     else if (diff > 0) i += diff
                  }

                  return (
                     <Button
                        onClick={() => handleButtonClick(i + 1)}
                        variant={i + 1 === page ? 'default' : 'outline'}
                        key={i}
                        className="px-2 lg:px-4 py-1 lg:py-2"
                        disabled={isFetching}
                     >
                        {i + 1}
                     </Button>
                  )
               })}
            </div>
            <button
               onClick={handleNextPage}
               disabled={page === Math.ceil(length / perPage)}
               className="text-3xl text-blue-500"
            >
               <BiChevronRight />
            </button>
            <button
               data-testid="next"
               onClick={() => navigate(`?page=${pages}`)}
               disabled={page === Math.ceil(length / perPage)}
               className="text-3xl text-blue-500"
            >
               <BiChevronsRight />
            </button>
         </div>
         <p className="text-center pt-3 font-proxima-medium">
            Всего страниц: {pages}
         </p>
      </>
   )
}

export default Pagination
