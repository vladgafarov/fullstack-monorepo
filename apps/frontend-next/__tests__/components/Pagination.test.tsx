import Pagination from '@components/common/Pagination'
import { render, screen, fireEvent } from '@lib/test-utils'
import { useRouter } from 'next/router'

describe('Pagination component test', () => {
   it('Should render', () => {
      render(<Pagination length={12} link="test" />)
   })

   it('Should be 4 pages', () => {
      const { container } = render(<Pagination length={12} link="test" />)

      expect(container).toHaveTextContent(/всего страниц: 4/i)
   })

   it('Should navigate to next page', () => {
      const router = useRouter()
      render(<Pagination length={12} link="test" />)

      const button = screen.getByRole('button', { name: /2/i })
      fireEvent.click(button)

      expect(router.push).toHaveBeenCalled()
      expect(router.push).toHaveBeenCalledWith('test?page=2')
   })

   it('Should navigate to prev page', () => {
      const router = useRouter()
      render(<Pagination length={12} link="test" />)

      const buttonNext = screen.getByRole('button', { name: /2/i })
      const buttonPrev = screen.getByRole('button', { name: /1/i })

      fireEvent.click(buttonNext)
      expect(router.push).toHaveBeenCalledWith('test?page=2')

      fireEvent.click(buttonPrev)
      expect(router.push).toHaveBeenCalledWith('test?page=1')
   })

   it('Should prev button disabled', () => {
      render(<Pagination length={12} link="test" />)

      const button = screen.getByTestId(/prev/i)

      expect(button).toBeDisabled()
   })

   it('Should next button disabled', () => {
      const router = useRouter()
      const { rerender } = render(<Pagination length={6} link="test" />)

      const button = screen.getByTestId(/next/i)
      fireEvent.click(button)

      expect(router.push).toHaveBeenCalledWith('test?page=2')

      router.query.page = '2'

      rerender(<Pagination length={6} link="test" />)
      expect(router.query.page).toEqual('2')

      expect(button).toBeDisabled()
   })

   it('Should not render when < 2 pages', () => {
      const { container } = render(<Pagination length={1} link="test" />)

      expect(container).toMatchSnapshot()
   })
})

export {}
