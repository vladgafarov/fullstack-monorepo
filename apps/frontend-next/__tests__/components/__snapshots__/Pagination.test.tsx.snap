// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component test Should not render when < 2 pages 1`] = `<div />`;
